cache results of `isomorphism`

ThomasBreuer opened the pull request:
as suggested by @fingolfin in #1243

ThomasBreuer committed:
cache results of `isomorphism`

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -390,11 +390,15 @@ Return an isomorphism from `G` to a group of type `T`.
-   return GAPGroupHomomorphism(G, H, f)
+   # Known isomorphism are cached in the attribute `:isomorphisms`.
+   isos = get_attribute!(() -> Dict{Type,GAPGroupHomomorphism}(), G, :isomorphisms)
+   return get!(isos, T) do
```
Perhaps we then should also do this for the variants for `GrpAbFinGen`?

Perhaps we should also document this caching behavior, so that people know they can rely on it?

ThomasBreuer replied:
Good idea. Thanks.

fingolfin replied:
On the long run it'd be nice if this was easier to do. Motivated me to file issue #1255

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -390,11 +390,15 @@ Return an isomorphism from `G` to a group of type `T`.
 An exception is thrown if no such isomorphism exists.
```
Perhaps we can also add an example/doctest to the docstring of this function. I wrote the patch below last night while working on PR #1246 but removed it again because I wanted that PR to focus. But perhaps we can add it (or a variant) now...

````suggestion
An exception is thrown if no such isomorphism exists.

If only the image of such an isomorphism is needed, use `T(G)`.

# Examples
```jldoctest
julia> G = dihedral_group(6)
<pc group of size 6 with 2 generators>

julia> iso = isomorphism(PermGroup, G)
Group homomorphism from
<pc group of size 6 with 2 generators>
to
Group([ (1,2)(3,6)(4,5), (1,3,5)(2,4,6) ])

julia> PermGroup(G)
Group([ (1,2)(3,6)(4,5), (1,3,5)(2,4,6) ])

julia> codomain(iso) == ans
true
```
````

Of course the final `codomain(iso) == ans` could be changed to `codomain(iso) === ans` with the changes in this PR.

ThomasBreuer replied:
Good idea. Thanks.

ThomasBreuer committed:
improvements

- document that isomorphisms are cached, and add a doctest to the docstring
- cache also isomorphisms involving `GrpAbFinGen` groups
- and added several type annotations in order to make `@inferred` happy
  (I do not understand why some of the tests started to fail right now because of missing type information.)

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -44,7 +44,7 @@ function symmetric_group(::Type{T}, n::Int) where T <: GAPGroup
     throw(ArgumentError("n must be a positive integer"))
   end
-  return T(GAP.Globals.SymmetricGroup(_gap_filter(T), n))
+  return T(GAP.Globals.SymmetricGroup(_gap_filter(T), n))::T
```
This should not be necessary resp. propably fixes the wrong problem. Better to use:
```suggestion
  return T(GAP.Globals.SymmetricGroup(_gap_filter(T), n)::GapObj)
```
This way Julia can prove which method for `T()` gets invoked, and thus correctly deduce the return type; this is better because now it has additional information, namely the `T` method being invoked, and can thus potentially perform further optimizations.

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -388,13 +388,40 @@ _get_iso_function(::Type{PcGroup}) = GAP.Globals.IsomorphismPcGroup
-   H = T(GAP.Globals.ImagesSource(f)::GapObj)
-   return GAPGroupHomomorphism(G, H, f)
+   # Known isomorphism are cached in the attribute `:isomorphisms`.
+   isos = get_attribute!(() -> Dict{Type, Any}(), G, :isomorphisms)
```
I think we fixed `get_attribute! to take type names, so perhaps try if this now works:
```suggestion
   isos = get_attribute!(Dict{Type, Any}, G, :isomorphisms)
```

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -388,13 +388,40 @@ _get_iso_function(::Type{PcGroup}) = GAP.Globals.IsomorphismPcGroup
+     f == GAP.Globals.fail && throw(ArgumentError("Could not convert group into a group of type $T"))
+     H = T(GAP.Globals.ImagesSource(f)::GapObj)
+     return GAPGroupHomomorphism(G, H, f)
+   end
```
```suggestion
   end::T
```

fingolfin replied:
OK this is wrong, but some type should be annotated

fingolfin replied:
Something like this
```suggestion
   end
   return result::GAPGroupHomomorphism{typeof(G),T}
```

fingolfin replied:
Maybe also this works (without the `result` helper), you can find out using `@code_warntype` and/or just trying the code.
```suggestion
   end::GAPGroupHomomorphism{typeof(G),T}
```

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -388,13 +388,40 @@ _get_iso_function(::Type{PcGroup}) = GAP.Globals.IsomorphismPcGroup
-   return GAPGroupHomomorphism(G, H, f)
+   # Known isomorphism are cached in the attribute `:isomorphisms`.
+   isos = get_attribute!(() -> Dict{Type, Any}(), G, :isomorphisms)
+   return get!(isos, T) do
```
```suggestion
   result = get!(isos, T) do
```

ThomasBreuer committed:
address comments

fingolfin committed:
More type stability
